src/category_theory/instances/Top/presheaf_of_functions.lean

```diff
@@ -0,0 +1,84 @@
+import category_theory.instances.Top.presheaf
+import category_theory.instances.TopCommRing.basic
+import category_theory.yoneda
+import group_theory.submonoid
+import ring_theory.subring
+import topology.algebra.continuous_functions
+
+universes v u
+
+open category_theory
+open category_theory.instances
+open topological_space
+
+open category_theory
+open category_theory.instances
+open topological_space
+
+namespace category_theory.instances.Top
+
+variables (X : Top.{v})
+
+def presheaf_to_Top (T : Top.{v}) : X.presheaf (Type v) :=
+(opens.to_Top X).op ⋙ (yoneda.obj T)
+
+-- TODO upgrade the result to TopCommRing?
+def continuous_functions (X : Top.{v}ᵒᵖ) (R : TopCommRing.{v}) : CommRing.{v} :=
```

Comment:
This seems rather specialised to commutative rings. I think this should be more abstract. We want the analogue for (abelian) groups, rings, monoids, etc... Either all these files get automatically generated, or we need to efficiently abstract away TopCommRing. Otherwise we'll get lots of files that go out of sync and it will be a nightmare to maintain.

Reply:
... I guess so. Can you think of the mathematical statement that will do this for us? I can't at the moment.

I'd argue that building examples of presheaves in CommRing is more important than presheaves of SemiBidistributiveNanoQuandles. :-)

Reply:
Really the thing that wants generalising is
```
def CommRing_yoneda : TopCommRing ⥤ (Topᵒᵖ ⥤ CommRing) :=
{ obj := λ R,
  { obj := λ X, continuous_functions X R,
    map := λ X Y f, continuous_functions.pullback f R },
  map := λ R S φ,
  { app := λ X, continuous_functions.map X φ } }
```
(perhaps even along with the fact that this preserves limits, so we get a sheaf, not just a presheaf?)

I'm not sure what name such generalisations go under, however. @rwbarton?

Reply:
I'm not sure how to properly generalize that, unless we want to define algebraic theories in general and connect those definitions with the definitions we already have, which seems like a lot of work.